rust/ql/lib/codeql/rust/internal/TypeMention.qll

```diff
@@ -329,6 +327,76 @@ class SelfTypeParameterMention extends TypeMention instanceof Name {
   }
 }
 
+/**
+ * Gets the type at `path` of the type being implemented in `i`, when
+ * `i` is an `impl` block, or the synthetic `Self` type parameter when
+ * `i` is a trait.
+ */
+pragma[nomagic]
+Type resolveImplOrTraitType(ImplOrTraitItemNode i, TypePath path) {
+  result = resolveImplSelfType(i, path)
+  or
+  result = TSelfTypeParameter(i) and path.isEmpty()
+}
+
+pragma[nomagic]
+private ImplOrTraitItemNode getSelfParamEnclosingImplOrTrait(SelfParam self) {
+  exists(Function f |
+    f = result.getAnAssocItem() and
+    self = f.getSelfParam()
+  )
+}
+
+/**
+ * An element used to represent the type of a `self` parameter that uses [shorthand
+ * syntax][1], which is sugar for an explicit annotation.
+ *
+ * [1]: https://doc.rust-lang.org/stable/reference/items/associated-items.html#r-associated.fn.method.self-pat-shorthands
+ */
+class ShorthandSelfParameterMention extends TypeMention instanceof SelfParam {
+  private ImplOrTraitItemNode encl;
+
+  ShorthandSelfParameterMention() {
+    not super.hasTypeRepr() and
+    encl = getSelfParamEnclosingImplOrTrait(this) and
+    (
+      encl instanceof Trait
+      or
+      // avoid generating a type mention if the type being implemented cannot be resolved
```

Comment:
I think it's hard to see how the comment and the code here fit together here.

What about changing the comment to:
```suggestion
      // avoid generating a type mention if the type being implemented does not have a type mention
```